File: domain/src/main/java/com/example/domain/usecase/project/assets/RemoveProjectProfileImageUseCase.kt
package com.example.domain.usecase.project.assets

import com.example.core_common.result.CustomResult
import com.example.domain.model.vo.DocumentId
import com.example.domain.repository.base.ProjectRepository
import javax.inject.Inject

/**
 * 프로젝트 프로필 이미지를 제거하는 UseCase 인터페이스
 */
interface RemoveProjectProfileImageUseCase {
    /**
     * 프로젝트의 프로필 이미지를 제거합니다.
     *
     * @param projectId 프로젝트 ID
     * @return 성공 시 성공 메시지가 포함된 CustomResult.Success, 실패 시 Exception이 포함된 CustomResult.Failure
     */
    suspend operator fun invoke(projectId: DocumentId): CustomResult<String, Exception>
}

/**
 * 프로젝트 프로필 이미지를 제거하는 UseCase 구현체
 *
 * @property projectRepository 프로젝트 정보 관련 기능을 제공하는 Repository
 */
class RemoveProjectProfileImageUseCaseImpl @Inject constructor(
    private val projectRepository: ProjectRepository
) : RemoveProjectProfileImageUseCase {
    
    /**
     * 프로젝트의 프로필 이미지를 제거합니다.
     * Firebase Functions를 통해 서버에서 프로젝트 프로필 이미지를 삭제합니다.
     *
     * @param projectId 프로젝트 ID
     * @return 성공 시 성공 메시지가 포함된 CustomResult.Success, 실패 시 Exception이 포함된 CustomResult.Failure
     */
    override suspend operator fun invoke(projectId: DocumentId): CustomResult<String, Exception> {
        return try {
            // Firebase Functions를 통해 프로젝트 프로필 이미지 삭제
            val result = projectRepository.removeProfileImage(projectId)
            
            when (result) {
                is CustomResult.Success -> {
                    CustomResult.Success("Project profile image removed successfully")
                }
                is CustomResult.Failure -> {
                    CustomResult.Failure(result.error)
                }
                else -> {
                    CustomResult.Failure(Exception("Unknown error occurred during project profile image removal"))
                }
            }
        } catch (e: Exception) {
            CustomResult.Failure(e)
        }
    }
}

File: functions/src/business/project/usecases/removeProjectProfileImage.usecase.ts
import { ProjectRepository } from '../../../domain/project/repositories/project.repository';
import { CustomResult, Result } from '../../../core/types';
import { NotFoundError, ValidationError } from '../../../core/errors';
import { FirebaseStorageService } from '../../../infrastructure/storage/firebase-storage.service';

export interface RemoveProjectProfileImageRequest {
  projectId: string;
  userId: string; // 요청하는 사용자 ID (권한 검증용)
}

export interface RemoveProjectProfileImageResponse {
  success: boolean;
  message: string;
}

export class RemoveProjectProfileImageUseCase {
  constructor(
    private readonly projectRepository: ProjectRepository,
    private readonly storageService: FirebaseStorageService
  ) {}

  async execute(request: RemoveProjectProfileImageRequest): Promise<CustomResult<RemoveProjectProfileImageResponse>> {
    try {
      // 1. 프로젝트 존재 확인
      const projectResult = await this.projectRepository.findById(request.projectId);
      if (!projectResult.success) {
        return Result.failure(projectResult.error);
      }

      const project = projectResult.data;
      if (!project) {
        return Result.failure(new NotFoundError('Project', request.projectId));
      }

      // 2. 사용자 권한 확인 (프로젝트 멤버인지 확인)
      const hasPermission = await this.checkUserPermission(request.projectId, request.userId);
      if (!hasPermission.success) {
        return Result.failure(hasPermission.error);
      }

      // 3. 프로젝트 이미지 경로 정의
      const profileImagePath = `project_profiles/${request.projectId}/profile.webp`;

      try {
        // 4. Firebase Storage에서 프로젝트 이미지 삭제
        await this.storageService.deleteFile(profileImagePath);
        
        // 5. 성공 응답
        return Result.success({
          success: true,
          message: 'Project profile image removed successfully'
        });
      } catch (storageError) {
        // 파일이 존재하지 않는 경우도 성공으로 간주
        if (this.isFileNotFoundError(storageError)) {
          return Result.success({
            success: true,
            message: 'Project profile image already removed or does not exist'
          });
        }
        
        // 다른 스토리지 에러는 실패로 처리
        return Result.failure(new Error(`Failed to delete project profile image: ${storageError instanceof Error ? storageError.message : 'Unknown storage error'}`));
      }
    } catch (error) {
      return Result.failure(new Error(`Failed to remove project profile image: ${error instanceof Error ? error.message : 'Unknown error'}`));
    }
  }

  private async checkUserPermission(projectId: string, userId: string): Promise<CustomResult<void>> {
    try {
      // 프로젝트 멤버인지 확인
      const memberResult = await this.projectRepository.findProjectMember(projectId, userId);
      if (!memberResult.success || !memberResult.data) {
        return Result.failure(new ValidationError('permission', 'User is not a member of this project'));
      }

      return Result.success(undefined);
    } catch (error) {
      return Result.failure(new Error(`Permission check failed: ${error instanceof Error ? error.message : 'Unknown error'}`));
    }
  }

  private isFileNotFoundError(error: any): boolean {
    return error?.code === 'storage/object-not-found' || 
           error?.message?.includes('No such object') ||
           error?.message?.includes('Object does not exist');
  }
}